mpz-common/src/sync/mod.rs

```diff
@@ -0,0 +1,293 @@
+///
+/// This is useful for when multiple different functionalities need to access some shared stateful resource.
+/// In such cases, it is important that the order in which this occurs is synchronized from all parties views.
+///
+/// [`Syncer`] employs a leader-follower coordination pattern, where the leader is responsible for
+/// dictating the order of execution, and the followers ensure their local execution is synchronized to it.
+///
+/// Every operation involves broadcasting a 4-byte ticket to all followers, so be conscious of the overhead
+/// of this.
+///
+/// See [`Mutex`] for a higher-level primitive that can be used to obtain a lock on a resource in synchronized
+/// order.
+#[derive(Debug, Clone)]
+pub struct Syncer(SyncerInner);
+
+impl Syncer {
+    /// Creates a new leader.
+    pub fn new_leader() -> Self {
+        Self(SyncerInner::Leader(Leader::default()))
+    }
+
+    /// Creates a new follower.
+    pub fn new_follower() -> Self {
+        Self(SyncerInner::Follower(Follower::default()))
+    }
+
+    /// Synchronizes the order of execution across logical threads.
+    ///
+    /// # Arguments
+    ///
+    /// * `io` - The I/O channel of the logical thread.
+    /// * `f` - The function to execute.
+    pub async fn sync<Io: IoDuplex<Ticket> + Unpin, F, R>(
+        &self,
+        io: &mut Io,
+        f: F,
+    ) -> Result<R, SyncError>
+    where
+        F: FnOnce() -> R + Unpin,
+        R: Unpin,
+    {
+        match &self.0 {
+            SyncerInner::Leader(leader) => leader.sync(io, f).await,
+            SyncerInner::Follower(follower) => follower.sync(io, f).await,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+enum SyncerInner {
+    Leader(Leader),
+    Follower(Follower),
+}
+
+#[derive(Debug, Default, Clone)]
+struct Leader {
+    tick: Arc<StdMutex<Ticket>>,
+}
+
+impl Leader {
+    async fn sync<Io: IoDuplex<Ticket> + Unpin, F, R>(
+        &self,
+        io: &mut Io,
+        f: F,
+    ) -> Result<R, SyncError>
+    where
+        F: FnOnce() -> R + Unpin,
+        R: Unpin,
+    {
+        let mut io = Pin::new(io);
+        poll_fn(|cx| io.as_mut().poll_ready(cx)).await?;
+        let (output, tick) = {
+            let mut tick_lock = self.tick.lock().unwrap();
+            let output = f();
+            let tick = tick_lock.increment_in_place();
+            (output, tick)
+        };
+        io.start_send(tick)?;
+        Ok(output)
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+struct Follower {
+    queue: Arc<StdMutex<Queue>>,
+}
+
+impl Follower {
+    async fn sync<Io: IoDuplex<Ticket> + Unpin, F, R>(
+        &self,
+        io: &mut Io,
+        f: F,
+    ) -> Result<R, SyncError>
+    where
+        F: FnOnce() -> R + Unpin,
+        R: Unpin,
+    {
+        let tick = io.expect_next().await?;
+        Ok(Wait::new(&self.queue, tick, f).await)
+    }
+}
+
+#[derive(Debug, Default)]
+struct Queue {
+    tick: Ticket,
```

Comment:
This being a ticket did not click for me. It is the current head of the queue, right? Can we somehow add this context information?